wgengine/netstack: TestTCPForwardLimits_PerClient data race in envknob

https://github.com/tailscale/tailscale/actions/runs/25776396947/job/75709875182?pr=19639
```
=== RUN   TestTCPForwardLimits_PerClient
==================
WARNING: DATA RACE
Write at 0x00c0001a06ef by goroutine 911:
  tailscale.com/envknob.setBoolLocked()
      /home/runner/work/tailscale/tailscale/src/envknob/envknob.go:219 +0x144
  tailscale.com/envknob.Setenv()
      /home/runner/work/tailscale/tailscale/src/envknob/envknob.go:103 +0x296
  tailscale.com/wgengine/netstack.TestTCPForwardLimits_PerClient()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack_test.go:817 +0x76
  testing.tRunner()
      /home/runner/.cache/tsgo/e877d973840c91ec9d4bc1921b0845789de359ae/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /home/runner/.cache/tsgo/e877d973840c91ec9d4bc1921b0845789de359ae/src/testing/testing.go:2101 +0x38

Previous read at 0x00c0001a06ef by goroutine 900:
  tailscale.com/envknob.RegisterBool.func1()
      /home/runner/work/tailscale/tailscale/src/envknob/envknob.go:154 +0x30
  tailscale.com/wgengine/netstack.(*Impl).Start.(*Impl).wrapTCPProtocolHandler.func1()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:561 +0x361
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverTransportPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/stack/nic.go:881 +0x576
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).deliverPacketLocally()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:1394 +0x113e
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).handleValidatedPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:1208 +0xe24
  gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).HandlePacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/network/ipv4/ipv4.go:931 +0x77e
  gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverNetworkPacket()
      /home/runner/work/tailscale/tailscale/gomodcache/gvisor.dev/gvisor@v0.0.0-20260224225140-573d5e7127a8/pkg/tcpip/stack/nic.go:781 +0x20d
  tailscale.com/wgengine/netstack.(*linkEndpoint).DeliverLoopback()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/link_endpoint.go:234 +0x2d4
  tailscale.com/wgengine/netstack.(*Impl).inject()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:1081 +0x11c
  tailscale.com/wgengine/netstack.(*Impl).Start.gowrap1()
      /home/runner/work/tailscale/tailscale/src/wgengine/netstack/netstack.go:647 +0x2e

Goroutine 911 (running) created at:
  testing.(*T).Run()
      /home/runner/.cache/tsgo/e877d973840c91ec9d4bc1921b0845789de359ae/src/testing/testing.go:2101 +0xb12
  testing.runTests.func1()
      /home/runner/.cache/tsgo/e877d973840c91ec9d4bc1921b0845789de359ae/src/testing/testing.go:2585 +0x84
  testing.tRunner()
      /home/runner/.cache/tsgo/e877d973840c91ec9d4bc1921b0845789de359ae/src/testing/testing.go:2036 +0x21c
  testing.runTests()
    logger.go:105:     ... control: [v1] HostInfo: {"IPNVersion":"1.99.0-ERR-BuildInfo","BackendLogID":"0000000000000000000000000000000000000000000000000000000000000000","OS":"linux","OSVersion":"6.17.0-1010-azure","Container":false,"Distro":"ubuntu","DistroVersion":"24.04","DistroCodeName":"noble","Desktop":false,"Hostname":"runnervmeorf1","Machine":"x86_64","GoArch":"amd64","GoArchVar":"v1","GoVersion":"go1.26.3","RoutableIPs":["192.0.2.0/24"],"Services":[{"Proto":"peerapi-dns-proxy","Port":1}],"Cloud":"azure","Userspace":true,"UserspaceRouter":true,"AppConnector":false,"StateEncrypted":false}
    local.go:2853:     ... Backend: logs: be:0000000000000000000000000000000000000000000000000000000000000000 fe:
    local.go:1119:     ... health(warnable=is-using-unstable-version): error: This is an unstable version of Tailscale meant for testing and development purposes. Please report any issues to Tailscale.
    wglog.go:66:     ... wg: [v2] Routine: receive incoming receiveDERP - started
    logger.go:105:     ... control: [v1] authRoutine: loggedIn=false; goal=nil paused=false
    local.go:6090:     ... Switching ipn state NoState -> NeedsLogin (WantRunning=false, nm=false)
    local.go:5228:     ... blockEngineUpdates(true)
    logger.go:105:     ... control: [v1] mapRoutine: loggedIn=false
    peermtu.go:95:     ... [v1] magicsock: peermtu: peer MTU status is false
    userspace.go:978:     ... wgengine: Reconfig: configuring router
    router_fake.go:26:     ... [v1] warning: fakeRouter.Set: not implemented.
    userspace.go:998:     ... wgengine: Reconfig: user dialer
    userspace.go:1007:     ... wgengine: Reconfig: configuring DNS
    logger.go:105:     ... dns: Set: {DefaultResolvers:[] Routes:{} SearchDomains:[] Hosts:0}
    logger.go:105:     ... dns: Resolvercfg: {Routes:{} Hosts:0 LocalDomains:[], cloud="azure"}
    logger.go:105:     ... dns: OScfg: {}
    userspace.go:1046:     ... [v1] wgengine: Reconfig done
    netstack.go:562:     ... [v2] netstack: in-flight connections for client 100.101.102.103: 0
    local.go:1119:     ... health(warnable=wantrunning-false): error: Tailscale is stopped.
    netstack.go:1511:     ... [v2] TCP ForwarderRequest: 100.101.102.103:1234 -> 192.0.2.1:4567
    netstack.go:1702:     ... [v2] netstack: forwarding incoming connection to 192.0.2.1:4567
    netstack_test.go:719: hangDialer: called with network="tcp" address="192.0.2.1:4567"
    netstack_test.go:874: got connection in progress
    netstack.go:562:     ... [v2] netstack: in-flight connections for client 100.101.102.103: 1
    netstack.go:565:     ... netstack: ignoring a new TCP connection from 192.0.2.1 to 100.101.102.103 because the client already has 1 in-flight connections
    netstack.go:562:     ... [v2] netstack: in-flight connections for client 192.0.2.1: 0
    netstack.go:589:     ... netstack: decrementing connsInFlightByClient[192.0.2.1] because the packet was not handled; new value is 0
    netstack_test.go:730: function completed
    netstack.go:1739:     ... netstack: could not connect to local backend server at 192.0.2.1:4567: canceled
    netstack.go:562:     ... [v2] netstack: in-flight connections for client 192.0.2.1: 0
    netstack.go:589:     ... netstack: decrementing connsInFlightByClient[192.0.2.1] because the packet was not handled; new value is 0
    logger.go:105:     ... control: client.Shutdown ...
    logger.go:105:     ... control: updateRoutine: exiting
    logger.go:105:     ... control: [v1] authRoutine: context done.
    logger.go:105:     ... control: authRoutine: exiting
    logger.go:105:     ... control: [v1] mapRoutine: context done.
    logger.go:105:     ... control: mapRoutine: exiting
    logger.go:105:     ... control: Client.Shutdown done.
    logger.go:105:     ... ipnext: work queue shutdown failed: execqueue shut down
    wglog.go:66:     ... wg: [v2] Routine: receive incoming mkReceiveFunc - stopped
    router_fake.go:31:     ... [v1] warning: fakeRouter.Close: not implemented.
    wglog.go:66:     ... wg: [v2] Device closing
    wglog.go:66:     ... wg: [v2] Routine: receive incoming mkReceiveFunc - stopped
    wglog.go:66:     ... wg: [v2] Routine: receive incoming receiveDERP - stopped
    wglog.go:66:     ... wg: [v2] Device closed
    testing.go:1712: race detected during execution of test
--- FAIL: TestTCPForwardLimits_PerClient (0.06s)
FAIL
FAIL	tailscale.com/wgengine/netstack	6.238s
```

--- a/wgengine/netstack/netstack.go
+++ b/wgengine/netstack/netstack.go
@@ -216,6 +216,7 @@ type Impl struct {
 	dialer    *tsdial.Dialer
 	ctx       context.Context        // alive until Close
 	ctxCancel context.CancelFunc     // called on Close
+	injectWG  sync.WaitGroup         // wait for the inject goroutine
 	lb        *ipnlocal.LocalBackend // or nil
 	dns       *dns.Manager
 
@@ -450,6 +451,7 @@ func (ns *Impl) Close() error {
 	ns.ctxCancel()
 	ns.ipstack.Close()
 	ns.ipstack.Wait()
+	ns.injectWG.Wait()
 	return nil
 }
 
@@ -644,7 +646,9 @@ func (ns *Impl) Start(b LocalBackend) error {
 	udpFwd := udp.NewForwarder(ns.ipstack, ns.acceptUDPNoICMP)
 	ns.ipstack.SetTransportProtocolHandler(tcp.ProtocolNumber, ns.wrapTCPProtocolHandler(tcpFwd.HandlePacket))
 	ns.ipstack.SetTransportProtocolHandler(udp.ProtocolNumber, ns.wrapUDPProtocolHandler(udpFwd.HandlePacket))
-	go ns.inject()
+	ns.injectWG.Go(func() {
+		ns.inject()
+	})
 	if ns.ready.Swap(true) {
 		panic("already started")
 	}

--- a/wgengine/netstack/netstack_test.go
+++ b/wgengine/netstack/netstack_test.go
@@ -737,7 +737,10 @@ func makeHangDialer(tb testing.TB) (netx.DialFunc, chan struct{}) {
 // TestTCPForwardLimits verifies that the limits on the TCP forwarder work in a
 // success case (i.e. when we don't hit the limit).
 func TestTCPForwardLimits(t *testing.T) {
+	tstest.AssertNotParallel(t) // calls envknob.Setenv
 	envknob.Setenv("TS_DEBUG_NETSTACK", "true")
+	t.Cleanup(func() { envknob.Setenv("TS_DEBUG_NETSTACK", "") })
+
 	impl := makeNetstack(t, func(impl *Impl) {
 		impl.ProcessSubnets = true
 	})
@@ -815,6 +818,7 @@ func TestTCPForwardLimits_PerClient(t *testing.T) {
 	clientmetric.ResetForTest(t)
 	tstest.AssertNotParallel(t) // calls envknob.Setenv
 	envknob.Setenv("TS_DEBUG_NETSTACK", "true")
+	t.Cleanup(func() { envknob.Setenv("TS_DEBUG_NETSTACK", "") })
 
 	// Set our test override limits during this test.
 	maxInFlightConnectionAttemptsForTest.Store(2)

